src/coreclr/jit/gentree.cpp

```diff
@@ -24313,6 +24317,29 @@ bool GenTreeHWIntrinsic::OperIsMemoryLoadOrStore() const
     return OperIsMemoryLoad() || OperIsMemoryStore();
 }
 
+//------------------------------------------------------------------------------
+// OperRequiresAsgFlag : Check whether the operation requires GTF_ASG flag regardless
+//                       of the children's flags.
+//
+bool GenTreeHWIntrinsic::OperRequiresAsgFlag() const
+{
+    if (OperIsMemoryStore())
+    {
+        // A MemoryStore operation is an assignment
+        return true;
+    }
+#if defined(TARGET_XARCH)
+    else if (HWIntrinsicInfo::HasSpecialSideEffect_Barrier(GetHWIntrinsicId()))
+    {
+        // While these don't technically do an assignment, they are modeled the
+        // same as GT_MEMORYBARRIER which tracks itself as requiring GT_ASG flag
+        return true;
+    }
+#endif // TARGET_XARCH
```

Comment:
```suggestion
#if defined(TARGET_XARCH)
    else if (HWIntrinsicInfo::HasSpecialSideEffect_Barrier(GetHWIntrinsicId()))
    {
        // We model memory barriers as opaque stores to globally visible memory.
        return true;
    }
#endif // TARGET_XARCH
```
There are quite a few cases where this pattern appears, can we add `OperIsMemoryStoreOrBarrier` to consolidate them?

Reply:
Maybe. I'll look at adding a helper, but not sure it actually helps in practice.